### PR TITLE
Implement reverse in libcudf

### LINKED ISSUE #7967: [FEA] Request a cudf function for `reverse`
**Is your feature request related to a problem? Please describe.**
I would be useful to have the ability to make a reverse copy of a column and/or table. 
I am thinking of `std::reverse` (https://en.cppreference.com/w/cpp/algorithm/reverse) . But not necessarily using iterators, since that might make things more complicated.

**Describe the solution you'd like**
The following APIs could be very useful (and hopefully self describing):
std::unique_ptr<column> reverse(column_view const& source);
std::unique_ptr<table> reverse(table_view const& source);
void reverse_in_place(mutable_column_view & mutable);
void reverse_in_place(mutable_table_view & mutable);

**Describe alternatives you've considered**
Right now if i want to make a reverse copy, i would have to make a `sequence` of indexed going from high to low and then go a `gather`, which is a lot less efficient.

**Additional context**

### PR DESCRIPTION
**Pull request by shaneding**
Closes #7967

### CODE REVIEW

File: cpp/include/cudf/copying.hpp [hunk 1]
```diff
@@ -81,6 +81,18 @@ std::unique_ptr<table> gather(
...
+/**
+ * @brief Reverses the rows within a table.
+ *
+ */
+std::unique_ptr<table> reverse(
+  table_view const& source_table,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+std::unique_ptr<column> reverse(
+  column_view const& source_table,
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp:93]:
```suggestion
  column_view const& source_column,
```

File: cpp/include/cudf/detail/reverse.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * @copydoc cudf::reverse(table_view const&,
+ * rmm::mr::device_memory_resource*)
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ */
+std::unique_ptr<table> reverse(
+  table_view const& source_table,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```
Comment (harrism) [inline cpp/include/cudf/detail/reverse.hpp:36]: Detail APIs should take streams (if they do anything that can run on a stream).
```suggestion
std::unique_ptr<table> reverse(
  table_view const& source_table,
  rmm::cuda_stream_view stream = rmm::cuda_stream_view{},
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

**Comment by harrism** [comment]
To pass the style failure, you just need to add reverse.hpp to the file `cudf/conda/recipes/libcudf/meta.yaml` (please add it in alphabetical order).

```
<     - test -f $PREFIX/include/cudf/detail/reverse.hpp
```

File: cpp/include/cudf/detail/reverse.hpp (hunk 2 above)
Comment (shaneding) [reply cpp/include/cudf/detail/reverse.hpp:36]: Should this be `rmm::cuda_stream_view stream = rmm::cuda_stream_default`? This seems to be the case in `repeat.hpp`.

Comment (jrhemstad) [reply cpp/include/cudf/detail/reverse.hpp]: > Should this be `rmm::cuda_stream_view stream = rmm::cuda_stream_default`? This seems to be the case in `repeat.hpp`.

A default constructed `cuda_stream_view` is the same as `cuda_stream_default`, so it doesn't really matter which you use.

File: cpp/include/cudf/copying.hpp [hunk 3]
```diff
@@ -81,6 +81,18 @@ std::unique_ptr<table> gather(
...
 
+/**
+ * @brief Reverses the rows within a table.
+ *
+ */
+std::unique_ptr<table> reverse(
+  table_view const& source_table,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+std::unique_ptr<column> reverse(
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp:92]: Need docs on this function too.

File: cpp/include/cudf/detail/reverse.hpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
```
Comment (harrism) [inline cpp/include/cudf/detail/reverse.hpp:2]: New file, so copyright just current year.  I think you are not obliged to use NVIDIA's copyright, just the rest of the license.
```suggestion
 * Copyright (c) 2021, NVIDIA CORPORATION.
```

**Comment by shaneding** [comment]
This is a relatively easy change, but I am unsure of what the desired return result should be from `reverse`. Under my current implementation. This is the result when calling `.reverse()` on a `cudf` dataframe.
```
>>> import cudf
>>> df = cudf.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
>>> df.reverse()
   a  b
0  3  6
1  2  5
2  1  4
```
However, the desired result in certain cases can also be:
```
>>> import cudf
>>> df = cudf.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
>>> df.reverse()
   a  b
2  3  6
1  2  5
0  1  4
```
Wherein the index is also reversed. Either way, it is a simple change. There is currently no equivalent of `.reverse()` inside pandas. An alternative could be to have a `reverse_index=False` parameter provided to `.reverse()`. 

Wanted to get some other people's opinions on this issue.
cc. @harrism @jrhemstad

**Comment by shwina** [comment]
From the Python side, this would be exposed not as a `reverse()` method, but rather using `iloc`:

```python
df.iloc[5:0-1, :] # get rows 0-5 in reverse order, all columns
```

I believe that will also reverse the index in Pandas.

**Comment by shwina** [comment]
It's totally fine if we tackled that in a future PR.

**Comment by shaneding** [comment]
> It's totally fine if we tackled that in a future PR.

Yeah I think this would be best tackled later in another PR. But if we are mimicking the behaviour of `.iloc[::-1]`, we should probably reverse the index here too then?

**Comment by shwina** [comment]
> we should probably reverse the index here too then?

Yup - agree.

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/include/cudf/detail/reverse.hpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (harrism) [inline cpp/include/cudf/detail/reverse.hpp:1]: I think unless and until the detail APIs are used outside of the implementation (e.g. by another translation unit), they should be kept inside the implementation .cu file. So no need for this header yet.

File: cpp/src/copying/reverse.cu [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+std::unique_ptr<column> reverse(column_view const& source_column,
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  table_view single_column                     = table_view({source_column});
+  auto reversed_column                         = std::move(cudf::reverse(single_column));
+  std::vector<std::unique_ptr<column>> columns = reversed_column->release();
+
+  return std::move(columns.at(0));
```
Comment (harrism) [inline cpp/src/copying/reverse.cu:58]: I think there's excessive `move`ment here.  :)

```suggestion
  return std::move(cudf::reverse({source_column})->release().front());
```

Note the implicit construction of the `table_view` from an init_list of `column_view`.  At least I think that should work.

File: cpp/tests/copying/reverse_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+template <typename T>
+class ReverseTypedTestFixture : public cudf::test::BaseFixture {
+};
+
+TYPED_TEST_CASE(ReverseTypedTestFixture, cudf::test::FixedWidthTypes);
+TYPED_TEST(ReverseTypedTestFixture, ReverseTable)
+{
+  using T = TypeParam;
+  static_assert(cudf::is_fixed_width<T>() == true, "this code assumes fixed-width types.");
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:40]: Why limit to fixed-width? Gather works for non-fixed-width.

File: cpp/tests/copying/reverse_tests.cpp [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    thrust::make_counting_iterator(0), thrust::make_counting_iterator(0));
+
+  auto expected = cudf::test::strings_column_wrapper(input_values.begin(), input_values.end());
+
+  cudf::table_view input_table{{input}};
+  auto p_ret = cudf::reverse(input_table);
+
+  EXPECT_EQ(p_ret->num_columns(), 1);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(p_ret->view().column(0), expected);
+}
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:186]: clang-format should fix this for you automatically. Do you not have clang-format set up to format on save?

```suggestion
}

```

**Review by harrism — CHANGES_REQUESTED** [review]
Nice first PR! Some suggestions on the C++.

File: cpp/include/cudf/copying.hpp [hunk 9]
```diff
@@ -81,6 +81,36 @@ std::unique_ptr<table> gather(
...
+ * Creates a new table that is the reverse of @p source_table.
+ * Example:
+ * ```
+ * source = [4,5,6]
+ * return = [6,5,4]
+ * ```
+ *
+ * @param source_table Table that will be reversed
+ */
+std::unique_ptr<table> reverse(
```
[the remark this answers is not in the document]
Comment (shaneding) [reply cpp/include/cudf/copying.hpp:95]: In my opinion, `.reverse()` here might be okay. IIRC, for row-specific operations like for example getting the unique elements in each row of lists, it is usually done like `df['list_col'].list.function()`, like [here](https://github.com/rapidsai/cudf/blob/5b8895dc50a5daf99bafede67617e8ebd39d3af6/python/cudf/cudf/core/column/lists.py#L237)

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/copying.hpp:95]: That's Python. I'm thinking about libcudf, which doesn't support the syntax you mention.

[the remark this answers is not in the document]
Comment (shaneding) [reply cpp/include/cudf/copying.hpp:95]: Oh I see what you mean, maybe something more specific like `reverse_table()` and `reverse_column()` could work?

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/copying.hpp:95]: I'm not complaining about the current name. Just thinking ahead.  e.g. for strings column concatenation we have `concatenate`, `join_strings` and `join_list_elements`, all of which do related but different things. 

My point is, we should think ahead about what the best names are for
1. A function that reverses the order of rows in a column or table
2. A function that reverses the order of list elements in each list (row) of a lists column.
3. A function that reverses the order of characters in each string (row) in a strings column.
4. A function that reverses the order of columns in a table (I doubt we would provide this at the libcudf level, just putting it here for completeness)

I think that `cudf::reverse` is fine for #1.  Ideally, we could find a common name for #2 and #3.

[the remark this answers is not in the document]
Comment (shaneding) [reply cpp/include/cudf/copying.hpp:95]: Gotcha, if we want a common name for 2 and 3, maybe `reverse_row_content()` could work here.

File: cpp/tests/copying/reverse_tests.cpp (hunk 8 above)
Comment (shaneding) [reply cpp/tests/copying/reverse_tests.cpp:186]: Not at the moment, I am just running the command provided in the cudf `CONTRIBUTING` file. But I will set it up.

File: cpp/include/cudf/copying.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (shwina) [reply cpp/include/cudf/copying.hpp:95]: Don't namespaces provide the necessary disambiguation? `cudf::reverse` v/s `cudf::strings::reverse` v/s `cudf::lists::reverse`?

File: cpp/src/copying/reverse.cu [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                                rmm::cuda_stream_view stream,
+                                rmm::mr::device_memory_resource* mr)
+{
+  return std::move(cudf::reverse(table_view({source_column}))->release().front());
+}
+}  // namespace detail
+
+std::unique_ptr<table> reverse(table_view const& source_table, rmm::mr::device_memory_resource* mr)
+{
+  return detail::reverse(source_table, rmm::cuda_stream_default, mr);
```
Comment (jrhemstad) [inline cpp/src/copying/reverse.cu:58]:
```suggestion
  CUDF_FUNC_RANGE();
  return detail::reverse(source_table, rmm::cuda_stream_default, mr);
```

File: cpp/src/copying/reverse.cu [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+std::unique_ptr<column> reverse(column_view const& source_column,
+                                rmm::mr::device_memory_resource* mr)
+{
+  return detail::reverse(source_column, rmm::cuda_stream_default, mr);
```
Comment (jrhemstad) [inline cpp/src/copying/reverse.cu:64]:
```suggestion
  CUDF_FUNC_RANGE();
  return detail::reverse(source_column, rmm::cuda_stream_default, mr);
```

**Comment by hyperbolic2346** [comment]
This looks good to me. I thought that we might want a gather map version of this since we seem to be moving towards gather maps, but the map is so trivial I don't see the value for it.

**Review by hyperbolic2346 — APPROVED** [review]
Once the naming thing is sorted I think this is fine to go.

File: cpp/include/cudf/copying.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/copying.hpp:95]: Good point.

File: cpp/tests/copying/reverse_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_view.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+
+template <typename T>
+class ReverseTypedTestFixture : public cudf::test::BaseFixture {
+};
+
+TYPED_TEST_CASE(ReverseTypedTestFixture, cudf::test::FixedWidthTypes);
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:36]: Still only testing FixedWidthTypes. Why?

File: cpp/tests/copying/reverse_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+  auto const p_ret = cudf::reverse(input_table);
+
+  EXPECT_EQ(p_ret->num_columns(), 1);
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(p_ret->view().column(0), expected, true);
+}
+
+TYPED_TEST(ReverseTypedTestFixture, ReverseColumn)
+{
+  using T = TypeParam;
+  // static_assert(cudf::is_fixed_width<T>() == true, "this code assumes fixed-width types.");
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:64]: Remove these commented out lines.

**Review by harrism — CHANGES_REQUESTED** [review]

File: cpp/tests/copying/reverse_tests.cpp (hunk 12 above)
Comment (shaneding) [reply cpp/tests/copying/reverse_tests.cpp:36]: My bad, thought I removed all of them.

File: python/cudf/cudf/core/frame.py [hunk 14]
```diff
@@ -1622,6 +1622,13 @@ def _repeat(self, count):
         result._copy_type_metadata(self)
         return result
 
+    def _reverse(self):
+        result = self.__class__._from_table(libcudf.copying.reverse(self))
+        return result
+
+    def reverse(self):
```
Comment (shwina) [inline python/cudf/cudf/core/frame.py:1629]: We probably don't want a public `reverse()` method, as Pandas doesn't expose one.

Comment (shwina) [reply python/cudf/cudf/core/frame.py:1629]: Note that `Frame` methods are exposed to end-users,  as the public classes `Series` and `DataFrame` and `Index` inherit from it.

Comment (shaneding) [reply python/cudf/cudf/core/frame.py:1629]: Makes sense, I will remove the public method.

**Review by shwina — APPROVED** [review]

**Review by brandon-b-miller — APPROVED** [review]

File: cpp/tests/copying/reverse_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+  std::vector<bool> input_valids(num_values);
+  for (size_t i{0}; i < input_valids.size(); i++) { input_valids[i] = (i % 2) == 0 ? true : false; }
+
+  std::vector<T> expected_values;
+  std::vector<bool> expected_valids;
+  for (int i = num_values - 1; i > -1; i--) {
+    expected_values.push_back(cudf::test::make_type_param_scalar<T>(input_values[i]));
+    expected_valids.push_back(input_valids[i]);
+  }
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:95]: See, named algorithms make your intent very clear here, and prevent future errors modifying tricky backward indexing:

```suggestion
  std::vector<T> expected_values(input_values.size());
  thrust::host_vector<bool> expected_valids(input_valids.size());
  std::copy_backward(input_values.begin(), input_values.end(), expected_values.begin());
  std::copy_backward(input_valids.begin(), input_valids.end(), expected_valids.begin());
```

File: cpp/tests/copying/reverse_tests.cpp [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+TEST_F(ReverseStringTestFixture, ReverseNullable)
+{
+  constexpr cudf::size_type num_values{20};
+
+  std::vector<std::string> input_values(num_values);
+  std::vector<bool> input_valids(num_values);
+  for (size_t i{0}; i < num_values; i++) {
+    input_values[i] = "#" + std::to_string(i);
+    input_valids[i] = (i % 2) == 0 ? true : false;
+  }
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:138]: Replace `vector<bool>` and use an algorithm for these.

File: cpp/tests/copying/reverse_tests.cpp [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+
+  std::vector<std::string> expected_values;
+  std::vector<bool> expected_valids;
+  for (int i = num_values - 1; i > -1; i--) {
+    expected_values.push_back(input_values[i]);
+    expected_valids.push_back(input_valids[i]);
+  }
```
Comment (harrism) [inline cpp/tests/copying/reverse_tests.cpp:145]: Replace `vector<bool>` and use an algorithm for these.

**Review by harrism — CHANGES_REQUESTED** [review]
Sorry, just noticed some test loops and `vector<bool>` that can be cleaned up.  Regarding "no raw loops", recommend this talk: https://www.youtube.com/watch?v=qH6sSOr-yk8

File: cpp/tests/copying/reverse_tests.cpp (hunk 15 above)
Comment (shaneding) [reply cpp/tests/copying/reverse_tests.cpp:95]: Gotcha! This does look much cleaner.

**Review by harrism — APPROVED** [review]
Tests look great!

**Comment by quasiben** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Review by trxcllnt — APPROVED** [review]

**Comment by shaneding** [comment]
@gpucibot merge

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**